### PR TITLE
Add `response.parsed_body` to Integration Tests

### LINKED ISSUE #23594: Provide @response.parsed_body to work together with post ..., as: :json for testing
If you post JSON in testing, it's likely you want to test JSON coming back too. Providing @response.parsed_body will save everyone from having to do `JSON.parse @response.body` over and over again.

### PR DESCRIPTION
**Pull request by kaspth**
Follow up to https://github.com/rails/rails/pull/21671

Fixes https://github.com/rails/rails/issues/23594

### CODE REVIEW

**Comment by rafaelfranca** [comment]
:shipit:

**Comment by kaspth** [comment]
Passed on this earlier build, https://travis-ci.org/rails/rails/builds/108383081, before I force pushed some changelog changes.

File: actionpack/lib/action_dispatch/testing/integration.rb [hunk 1]
```diff
@@ -422,17 +424,21 @@ def encode_params(params)
             @param_encoder.call(params)
           end
 
+          def parse_body(body)
+            @response_parser.call(body)
+          end
```
Comment (dhh) [inline actionpack/lib/action_dispatch/testing/integration.rb]: Would be nice if this was memoized. Frequent to have a bunch of checks against response.parsed_body['key1'], response.parsed_body['key2'].

**Comment by dhh** [comment]
We should extend this to work for the get case as well and whenever format is set:

```ruby
  get circle_people_url(format: :json)
  assert_equal "something", response.parsed_body['key'] 
```

Right now it doesn't as format isn't used to detect the mime.

File: actionpack/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,41 @@
...
+
+    Can now be written as:
+
+    ```ruby
+    require 'test_helper'
+
+    class ApiTest < ActionDispatch::IntegrationTest
+      test 'creates articles' do
+        assert_difference -> { Article.count } do
+          post articles_path, { article: { title: 'Ahoy!' } }, as: :json
```
Comment (dhh) [inline actionpack/CHANGELOG.md]: This should be: `post articles_path, params: { article: { title: 'Ahoy!' } }, as: :json`

Comment (kaspth) [inline actionpack/CHANGELOG.md]: Doh! 😁

**Comment by kaspth** [comment]
@dhh ah, you'd like us to replace `as` with just `format: :json` on the URL? Don't know whether or not that introduces edge cases because we'd have to parse the returned URL string a bit.

**Comment by dhh** [comment]
I'd like for response.parsed_body to work if the response is encoded as
JSON. Shouldn't be dependent on as: :json.

**Comment by matthewd** [comment]
Yeah, this shouldn't need to be based on the request encoding at all: the response headers tell us whether the response is JSON.

**Comment by kaspth** [comment]
@matthewd ah, of course 👍. Will patch this up tonight.

**Comment by kaspth** [comment]
@dhh @matthewd fixed your suggestions here https://github.com/rails/rails/compare/f9f98b75b57f...036a7a03cf43

**Comment by dhh** [comment]
Super!

**Comment by kaspth** [comment]
@dhh wups! Just noticed this doesn't work with `format: :json` calls. I'll look into it.

**Comment by f3ndot** [comment]
I noticed this still doesn't work with RSpec unit testing controllers for v5.2.4.3:

```ruby
RSpec.describe Api::MyApi, type: :controller do
  controller do
    def create
      render json: { foo: 'bar' }
    end
  end

  it do
    @request.headers['Accept'] = 'application/json'
    @request.headers['Content-Type'] = 'application/json'
    post :create, {}, as: :json, format: json
    expect(response.parsed_body).to eq( 'foo' => 'bar' ) # fails
  end
end
```

Inspecting the code, it looks like this is because the `@request` is instantiated long before `post()` is invoked (during `ActionController::TestCase::Behavior#setup_controller_request_and_response`) where content-type is nil and thus the `IdentityEncoder` is used, not the json parsing lambda.

As far as I can see, additional work is required to change the `@response_parser` in TestResponse to match whatever the latest assigned content_type is on time of controller execution

**Pull request merged**